File: contract/go/compute.go
package contract

// ShardKind selects the parallelism strategy for a compute shard.
type ShardKind uint8

const (
	ShardPipeline ShardKind = iota
	ShardTensor
	ShardData
)

// Shard describes a slice of a model/computation.
type Shard struct {
	Kind             ShardKind
	LayerLo, LayerHi uint32
	TPRank, TPWorld  uint32
}

// Promise is an unresolved future referenced before its producer finishes.
type Promise struct {
	PromiseID []byte
	Producer  PeerID
}

// TensorDType selects the element type of an ActivationFrame payload.
type TensorDType uint8

const (
	TensorDTypeUnspecified TensorDType = iota
	TensorDTypeF32
	TensorDTypeF16
	TensorDTypeBF16
	TensorDTypeI8
)

// CompressionHint names compression on the data-plane activation path.
// CompressionFrontierZK is a documented stub (zk-compressed bundles are Frontier).
const (
	CompressionUnspecified CompressionHint = iota
	CompressionNone
	CompressionZstd
	CompressionLZ4
	CompressionFrontierZK
)

// CompressionHint names compression on the data-plane activation path.
type CompressionHint uint8

// ActivationFrame mirrors compute.proto — tensor bytes plus shape metadata.
// Bulk payload may also ride QUIC directly; this tags what was (or will be) sent.
type ActivationFrame struct {
	Payload     []byte
	Shape       []uint32
	DType       TensorDType
	Compression CompressionHint
	StageIndex  uint32
}

// PipelineStageMeta records one shard placement in a pipeline plan.
type PipelineStageMeta struct {
	StageIndex uint32
	Shard      Shard
	Node       PeerID
}

// ComputeTask mirrors compute.proto / ARCHITECTURE.md §3.4.
type ComputeTask struct {
	TaskID    []byte
	Component []byte // IPLD CID of the WASM component
	Shard     Shard
	Caps      [][]byte // capabilities granted to this task
	Deps      []Promise
	ResultCap []byte
	// Activation is the structured inbound tensor (optional; Caps[1] inline bytes
	// remain valid for v0.1 pipeline demos).
	Activation ActivationFrame
	// PipelineStage is this shard's index in the pipeline plan (optional).
	PipelineStage uint32
}

// InferenceTask is an explicit inference/pipeline dispatch view over ComputeTask.
type InferenceTask struct {
	Task          ComputeTask
	Activation    ActivationFrame
	PipelineStage uint32
	Stages        []PipelineStageMeta
}

// ToComputeTask projects an InferenceTask to the wire ComputeTask shape.
func (it InferenceTask) ToComputeTask() ComputeTask {
	t := it.Task
	if len(it.Activation.Payload) > 0 || len(it.Activation.Shape) > 0 {
		t.Activation = it.Activation
	}
	if it.PipelineStage != 0 {
		t.PipelineStage = it.PipelineStage
	}
	return t
}

// ComputeResult is returned for a completed task (the v0.1 demo carries Output).
type ComputeResult struct {
	TaskID []byte
	OK     bool
	Output []byte
	Error  string
}

// CrdtOp mirrors state.proto / ARCHITECTURE.md §3.3.
type CrdtOp struct {
	DocID  []byte
	Actor  PeerID
	Clock  VectorClock
	Domain string // kv | counter | set | log | agent.belief
	Delta  []byte
	Cap    []byte
	Sig    []byte
}

// BeliefConflict is emitted instead of a silent merge for domain "agent.belief".
type BeliefConflict struct {
	DocID      []byte
	Subject    string
	Candidates []BeliefCandidate
	DetectedAt uint64
}

type BeliefCandidate struct {
	Actor PeerID
	Value []byte
	Clock VectorClock
}

File: contract/rust/src/compute.rs
//! Compute / inference contract types mirroring `proto/cerberus/v1/compute.proto`.

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub enum ShardKind {
    #[default]
    Pipeline = 0,
    Tensor = 1,
    Data = 2,
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Shard {
    pub kind: ShardKind,
    pub layer_lo: u32,
    pub layer_hi: u32,
    pub tp_rank: u32,
    pub tp_world: u32,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Promise {
    pub promise_id: Vec<u8>,
    pub producer: super::PeerId,
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub enum TensorDType {
    #[default]
    Unspecified = 0,
    F32 = 1,
    F16 = 2,
    BF16 = 3,
    I8 = 4,
}

/// Compression on the activation data path. `FrontierZk` is a documented stub.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub enum CompressionHint {
    #[default]
    Unspecified = 0,
    None = 1,
    Zstd = 2,
    Lz4 = 3,
    /// Frontier: zk-compressed proof bundles — not implemented in v0.1.
    FrontierZk = 4,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActivationFrame {
    pub payload: Vec<u8>,
    pub shape: Vec<u32>,
    pub dtype: TensorDType,
    pub compression: CompressionHint,
    pub stage_index: u32,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct PipelineStageMeta {
    pub stage_index: u32,
    pub shard: Shard,
    pub node: super::PeerId,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ComputeTask {
    pub task_id: Vec<u8>,
    pub component: Vec<u8>,
    pub shard: Shard,
    pub caps: Vec<Vec<u8>>,
    pub deps: Vec<Promise>,
    pub result_cap: Vec<u8>,
    pub activation: ActivationFrame,
    pub pipeline_stage: u32,
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct InferenceTask {
    pub task: ComputeTask,
    pub activation: ActivationFrame,
    pub pipeline_stage: u32,
    pub stages: Vec<PipelineStageMeta>,
}

impl InferenceTask {
    pub fn to_compute_task(&self) -> ComputeTask {
        let mut t = self.task.clone();
        if !self.activation.payload.is_empty() || !self.activation.shape.is_empty() {
            t.activation = self.activation.clone();
        }
        if self.pipeline_stage != 0 {
            t.pipeline_stage = self.pipeline_stage;
        }
        t
    }
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ComputeResult {
    pub task_id: Vec<u8>,
    pub ok: bool,
    pub output: Vec<u8>,
    pub error: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn inference_task_projection() {
        let it = InferenceTask {
            task: ComputeTask {
                task_id: b"tid".to_vec(),
                ..Default::default()
            },
            activation: ActivationFrame {
                payload: vec![0; 4],
                shape: vec![1],
                dtype: TensorDType::F32,
                compression: CompressionHint::None,
                ..Default::default()
            },
            pipeline_stage: 1,
            ..Default::default()
        };
        let ct = it.to_compute_task();
        assert_eq!(ct.pipeline_stage, 1);
        assert_eq!(ct.activation.payload.len(), 4);
    }

    #[test]
    fn activation_frame_serde_roundtrip() {
        let frame = ActivationFrame {
            payload: vec![1, 2, 3, 4],
            shape: vec![1],
            dtype: TensorDType::F32,
            compression: CompressionHint::None,
            stage_index: 2,
        };
        let json = serde_json::to_string(&frame).unwrap();
        let got: ActivationFrame = serde_json::from_str(&json).unwrap();
        assert_eq!(got, frame);
    }
}
